File: Game/Android/app/src/main/java/com/example/chessclassic/game/board/Square.java
package com.example.chessclassic.game.board;

import com.example.chessclassic.game.board.piece.Piece;

public class Square {
    public int pos;
    public Piece piece;
    public Square(int pos){
        this.pos = pos;
        clear();
    }
    public Square(Piece piece, int pos){
        this.piece = piece.copy();
        this.piece.pos = pos;
        this.pos = pos;
    }

    public void clear(){
        piece = null;
    }
    public boolean isOccupied(){
        return piece != null;
    }
    public Square copy(){
        return new Square(piece, pos);
    }
}

File: cpp/game/board/square.cpp
#include "square.hpp"

Square::Square() {
    clear();
}
Square::Square(Piece* piece) {
    this->piece = piece->copy();
    this->piece->pos = piece->pos;
}

void Square::clear() {
    piece = NULL;
}
bool Square::isOccupied() {
    return piece == NULL;
}
Square* Square::copy() {
    return isOccupied() ? new Square(piece) : new Square();
}